# NumPy nível 2 — Turma 5

**Objetivo:** organizar dados em matrizes, resumir por linha ou coluna e combinar regras de seleção. Inclui **15 exercícios progressivos e uma oficina extra com 8 exercícios de broadcasting, todos com resoluções comentadas**.

## Como estudar

Você já deve conhecer `np.array`, índices, fatiamento, cálculos por elemento, filtros simples e `np.where`. Se precisar, consulte o notebook de nível 1.

Execute as células de cima para baixo. Leia, preveja a saída e só então execute. Faça sua tentativa antes de consultar as resoluções, que ficam ao final. Não precisamos baixar dados ou criar gráficos.

**Roteiro:** entender → identificar dados e regra → resolver à mão → planejar → executar → conferir.

Se a importação falhar, retire o `#` da instalação opcional abaixo e execute. Depois, se necessário, reinicie o kernel e execute novamente as células.


In [1]:
# %pip install numpy

In [2]:
import numpy as np
print("NumPy nível 2 pronto!")

NumPy nível 2 pronto!


## 1. Matriz: linhas, colunas e significado

Uma matriz é um array de duas dimensões. Vamos representar vendas: **cada linha é uma loja**; **cada coluna é um dia**.

| Loja | Segunda | Terça | Quarta |
|---|---|---|---|
| A | 10 | 20 | 30 |
| B | 40 | 50 | 60 |

O NumPy guarda os números, mas não guarda automaticamente esses nomes de lojas e dias. Precisamos saber o que cada posição representa.

In [3]:
vendas = np.array([[10, 20, 30], [40, 50, 60]])
print(vendas)
print("Formato:", vendas.shape)
print("Dimensões:", vendas.ndim)
print("Elementos:", vendas.size)

[[10 20 30]
 [40 50 60]]
Formato: (2, 3)
Dimensões: 2
Elementos: 6


`shape` é `(2, 3)`: duas linhas e três colunas. `ndim` é 2 e `size` é 6.

## 2. Selecionar linhas, colunas e blocos

`matriz[linhas, colunas]`: antes da vírgula selecionamos linhas; depois, colunas. `:` seleciona todas as posições naquela dimensão. Os índices começam em zero.

`1:3` seleciona os índices 1 e 2: o fim não entra.

In [4]:
print("Loja A:", vendas[0, :])
print("Terça de todas as lojas:", vendas[:, 1])
print("Terça e quarta das duas lojas:\n", vendas[:, 1:3])
print("Loja B na quarta:", vendas[1, 2])

Loja A: [10 20 30]
Terça de todas as lojas: [20 50]
Terça e quarta das duas lojas:
 [[20 30]
 [50 60]]
Loja B na quarta: 60


Os resultados são `[10 20 30]`, `[20 50]`, o bloco `[[20, 30], [50, 60]]` e o valor 60. Selecionar uma linha ou coluna com um único índice reduz uma dimensão neste exemplo.

## 3. Entender axis: por loja ou por dia

**`axis` indica a dimensão sobre a qual a operação é feita.** Em uma matriz, `axis=0` percorre a dimensão das linhas, juntando valores das linhas para cada coluna. `axis=1` percorre a dimensão das colunas, juntando os valores de cada linha.

| Pedido | Comando | Conta |
|---|---|---|
| Total geral | `vendas.sum()` | Todos os valores |
| Total de cada dia | `vendas.sum(axis=0)` | 10+40; 20+50; 30+60 |
| Total de cada loja | `vendas.sum(axis=1)` | 10+20+30; 40+50+60 |

`axis=0` e `axis=1` são **argumentos nomeados**: escrevemos o nome do parâmetro e o valor fornecido.

In [5]:
print("Total geral:", vendas.sum())
print("Total por dia:", vendas.sum(axis=0))
print("Total por loja:", vendas.sum(axis=1))
print("Média por dia:", vendas.mean(axis=0))
print("Média por loja:", vendas.mean(axis=1))

Total geral: 210
Total por dia: [50 70 90]
Total por loja: [ 60 150]
Média por dia: [25. 35. 45.]
Média por loja: [20. 50.]


O total geral é 210. Por dia, `[50 70 90]`; por loja, `[60 150]`. As médias são `[25. 35. 45.]` e `[20. 50.]`. Em duas dimensões, somar com um axis deixa os resultados organizados pela outra dimensão.

## 4. reshape: reorganizar o formato

`reshape` reorganiza os elementos em um novo formato. **A quantidade total de elementos deve continuar igual.** Se há seis elementos, podemos criar 2 × 3 ou 3 × 2; 2 × 4 não cabe.

O NumPy pode inferir um tamanho quando usamos `-1` em uma dimensão.

In [6]:
sequencia = np.arange(1, 7)
matriz_2x3 = sequencia.reshape(2, 3)
print("Sequência:", sequencia)
print("2 por 3:\n", matriz_2x3)
print("3 por 2:\n", sequencia.reshape(3, 2))
print("Duas linhas, colunas inferidas:\n", sequencia.reshape(2, -1))

Sequência: [1 2 3 4 5 6]
2 por 3:
 [[1 2 3]
 [4 5 6]]
3 por 2:
 [[1 2]
 [3 4]
 [5 6]]
Duas linhas, colunas inferidas:
 [[1 2 3]
 [4 5 6]]


`arange(1, 7)` cria 1 a 6. `reshape(2, 3)` organiza em `[[1, 2, 3], [4, 5, 6]]`. reshape pode compartilhar dados com o original; não o trate como cópia independente.

## 5. Transpor e voltar a uma sequência

`.T` é o atributo que fornece a transposição: troca os eixos de linhas e colunas. Não ordena os números.

`.flatten()` devolve uma cópia com uma dimensão, na ordem padrão de leitura por linhas.

In [7]:
print("Original:\n", matriz_2x3)
print("Transposta:\n", matriz_2x3.T)
print("Formato da transposta:", matriz_2x3.T.shape)
print("Uma dimensão:", matriz_2x3.flatten())

Original:
 [[1 2 3]
 [4 5 6]]
Transposta:
 [[1 4]
 [2 5]
 [3 6]]
Formato da transposta: (3, 2)
Uma dimensão: [1 2 3 4 5 6]


A transposta tem formato `(3, 2)` e valores `[[1, 4], [2, 5], [3, 6]]`. flatten devolve `[1 2 3 4 5 6]`. A transposição normalmente compartilha os dados; flatten cria uma cópia.

## 6. Broadcasting: combinar formatos compatíveis

**Broadcasting** permite calcular com arrays de formatos diferentes quando eles são compatíveis. Um número somado a um array é um caso simples que já usamos.

Para comparar formatos, alinhamos as dimensões pelo final. Cada par deve ter o mesmo tamanho ou um dos tamanhos deve ser 1. Dimensões ausentes são tratadas como tamanho 1.

Vamos somar um bônus diferente para cada dia: a matriz tem formato `(2, 3)`; o bônus tem `(3,)`. As três posições do bônus correspondem às três colunas.

In [8]:
bonus_dia = np.array([1, 2, 3])
vendas_com_bonus = vendas + bonus_dia
print(vendas_com_bonus)

# Um bônus por loja precisa estar em uma coluna: formato (2, 1).
bonus_loja = np.array([5, 10]).reshape(2, 1)
print("Bônus por loja:\n", vendas + bonus_loja)

[[11 22 33]
 [41 52 63]]
Bônus por loja:
 [[15 25 35]
 [50 60 70]]


O bônus por dia produz `[[11, 22, 33], [41, 52, 63]]`. O bônus por loja produz `[[15, 25, 35], [50, 60, 70]]`. Um array de formato `(2,)` não serviria para esse bônus por loja: sua dimensão final 2 conflita com as três colunas.

### Broadcasting com calma — uma regra para várias posições

**Broadcasting é a regra usada pelo NumPy para combinar formatos compatíveis em operações por elemento.** Ele aplica valores em várias posições sem exigir que você crie manualmente um array cheio de repetições.

Pense em uma papelaria: as **linhas são filiais** e as **colunas são produtos**. Antes de fazer qualquer conta, dê um significado a cada eixo.

| Filial | Caderno | Caneta | Pasta |
|---|---|---|---|
| A | 10 | 20 | 30 |
| B | 40 | 50 | 60 |

Esta matriz tem `shape (2, 3)`. A posição de um número importa: um aumento por produto é diferente de um aumento por filial.

### Caso 1 — um número para todos

Em `matriz + 5`, o mesmo 5 é somado a cada posição. Um número sozinho é chamado de **escalar** e tem formato `()` no NumPy.


In [9]:
base_b = np.array([[10, 20, 30], [40, 50, 60]])
print("Formato da base:", base_b.shape)
print("Formato do escalar:", np.asarray(5).shape)
print("Mais 5 em todas as posições:\n", base_b + 5)

Formato da base: (2, 3)
Formato do escalar: ()
Mais 5 em todas as posições:
 [[15 25 35]
 [45 55 65]]


O resultado é `[[15, 25, 35], [45, 55, 65]]`. Visualmente, imagine somar uma matriz de cincos; o NumPy não precisa criar essa matriz de repetições para aplicar a regra.

### Caso 2 — um valor por coluna

Um array `[1, 2, 3]` tem formato `(3,)`. Ele tem uma dimensão: **não tem orientação de linha ou coluna por si só**. Ao operar com uma matriz de formato `(2, 3)`, sua dimensão é alinhada à dimensão final da matriz: as três colunas.

| Filial | Caderno | Caneta | Pasta |
|---|---|---|---|
| A | 10 + 1 | 20 + 2 | 30 + 3 |
| B | 40 + 1 | 50 + 2 | 60 + 3 |


In [10]:
por_produto_b = np.array([1, 2, 3])
print("Formato do ajuste:", por_produto_b.shape)
print(base_b + por_produto_b)

Formato do ajuste: (3,)
[[11 22 33]
 [41 52 63]]


Resultado: `[[11, 22, 33], [41, 52, 63]]`. O ajuste 1 pertence à primeira coluna; 2 à segunda; 3 à terceira.

Também podemos escrever o ajuste com formato `(1, 3)`: **uma linha e três colunas**. O tamanho 1 na primeira dimensão permite que essa linha seja usada nas duas linhas da base.


In [11]:
linha_b = por_produto_b.reshape(1, 3)
print("Ajuste como linha:\n", linha_b)
print("Formato:", linha_b.shape)
print(base_b + linha_b)

Ajuste como linha:
 [[1 2 3]]
Formato: (1, 3)
[[11 22 33]
 [41 52 63]]


### Caso 3 — um valor por linha

Agora o ajuste é 5 para a filial A e 10 para a filial B. Precisamos de **duas linhas e uma coluna**: formato `(2, 1)`.

| Filial | Caderno | Caneta | Pasta |
|---|---|---|---|
| A | 10 + 5 | 20 + 5 | 30 + 5 |
| B | 40 + 10 | 50 + 10 | 60 + 10 |


In [12]:
por_filial_b = np.array([5, 10])
coluna_b = por_filial_b.reshape(2, 1)
print("Antes:", por_filial_b.shape)
print("Depois:", coluna_b.shape)
print("Ajuste como coluna:\n", coluna_b)
print("Resultado:\n", base_b + coluna_b)

Antes: (2,)
Depois: (2, 1)
Ajuste como coluna:
 [[ 5]
 [10]]
Resultado:
 [[15 25 35]
 [50 60 70]]


Resultado: `[[15, 25, 35], [50, 60, 70]]`. O tamanho 1 da última dimensão permite aplicar o ajuste daquela linha nas três colunas.

**Por que precisamos do reshape?** Um array de formato `(2,)` é alinhado às colunas ao operar com `(2, 3)`. Ele não se torna uma coluna automaticamente: 2 e 3 são incompatíveis.

### Regra de compatibilidade — comparar da direita para a esquerda

1. Consulte os formatos com `.shape`.
2. Alinhe as dimensões pelo final.
3. Se faltar uma dimensão, considere tamanho 1 nessa posição.
4. Cada par precisa ter tamanhos **iguais**, ou **um deles ser 1**.
5. Se um par não atender à regra, a operação falha.

O formato do resultado usa o tamanho compatível de cada par. Nos exemplos com tamanhos positivos abaixo, é o maior tamanho do par.

| Formatos | Comparação da dimensão final | Dimensão anterior | Funciona? | Resultado |
|---|---|---|---|---|
| `(2, 3)` e `()` | 3 e 1 | 2 e 1 | Sim | `(2, 3)` |
| `(2, 3)` e `(3,)` | 3 e 3 | 2 e 1 implícito | Sim | `(2, 3)` |
| `(2, 3)` e `(1, 3)` | 3 e 3 | 2 e 1 | Sim | `(2, 3)` |
| `(2, 3)` e `(2, 1)` | 3 e 1 | 2 e 2 | Sim | `(2, 3)` |
| `(2, 3)` e `(2,)` | 3 e 2 | Não precisamos continuar | Não | Erro |
| `(2, 3)` e `(3, 2)` | 3 e 2 | Não precisamos continuar | Não | Erro |

**Atenção:** ter o mesmo total de elementos não garante broadcasting. Os formatos `(2, 3)` e `(3, 2)` têm seis elementos cada, mas são incompatíveis nessa operação.


In [13]:
# O erro é intencional e está tratado para o notebook continuar.
try:
    resultado_invalido_b = base_b + np.array([5, 10])
except ValueError as erro:
    print("Não funciona:", erro)

# Correção: indicar explicitamente o ajuste por linha.
print("Corrigido:\n", base_b + np.array([5, 10]).reshape(2, 1))

Não funciona: operands could not be broadcast together with shapes (2,3) (2,) 
Corrigido:
 [[15 25 35]
 [50 60 70]]


`try` tenta executar o bloco. `except ValueError` trata o erro de formatos incompatíveis. Usamos isso apenas para observar um erro sem interromper o notebook.

### Caso 4 — coluna + linha: todas as combinações

Uma coluna de formato `(2, 1)` e uma linha de formato `(1, 3)` produzem uma matriz `(2, 3)`.

Considere dois valores de frete, 10 e 20, e três preços, 100, 200 e 300. Queremos calcular o preço com cada opção de frete.


In [14]:
fretes_b = np.array([10, 20]).reshape(2, 1)
precos_b = np.array([100, 200, 300]).reshape(1, 3)
combinacoes_b = fretes_b + precos_b
print("Fretes:", fretes_b.shape)
print("Preços:", precos_b.shape)
print("Todas as combinações:\n", combinacoes_b)
print("Resultado:", combinacoes_b.shape)

Fretes: (2, 1)
Preços: (1, 3)
Todas as combinações:
 [[110 210 310]
 [120 220 320]]
Resultado: (2, 3)


A primeira linha é `[110, 210, 310]`; a segunda, `[120, 220, 320]`. É uma operação por elemento depois da expansão compatível, não uma multiplicação de matrizes.

### Caso 5 — descontos diferentes por produto

Se cada coluna é um produto, um array de três fatores aplica um desconto diferente por coluna. Descontos de 10%, 20% e 0% correspondem aos fatores 0.90, 0.80 e 1.00.


In [15]:
precos_filiais_b = np.array([[100, 200, 50], [120, 180, 80]])
fatores_b = np.array([0.90, 0.80, 1.00])
print(precos_filiais_b * fatores_b)

[[ 90. 160.  50.]
 [108. 144.  80.]]


Os resultados são `[[90, 160, 50], [108, 144, 80]]`. Cada fator é usado nas duas filiais, na coluna correspondente.

### Caso 6 — subtrair a média de cada linha

Podemos retirar a média de cada loja para observar quanto cada dia está acima ou abaixo daquela média. `mean(axis=1)` devolve formato `(2,)`; para subtrair por linha, precisamos de `(2, 1)`.

**`keepdims=True` mantém a dimensão reduzida com tamanho 1**, evitando um reshape separado neste caso.


In [16]:
medias_b = base_b.mean(axis=1, keepdims=True)
print("Médias por linha:\n", medias_b)
print("Formato:", medias_b.shape)
print("Diferença para a média:\n", base_b - medias_b)

Médias por linha:
 [[20.]
 [50.]]
Formato: (2, 1)
Diferença para a média:
 [[-10.   0.  10.]
 [-10.   0.  10.]]


As médias são `[[20], [50]]`. O resultado é `[[-10, 0, 10], [-10, 0, 10]]`. Negativo significa abaixo da média; positivo, acima.

### Caso 7 — np.newaxis: inserir uma dimensão

`np.newaxis` pode inserir uma dimensão de tamanho 1. Neste exemplo, é outra forma de expressar linha ou coluna, sem mudar os valores.

| Expressão | Formato | Leitura |
|---|---|---|
| `a` | `(3,)` | Uma dimensão |
| `a[np.newaxis, :]` | `(1, 3)` | Insere dimensão antes: uma linha |
| `a[:, np.newaxis]` | `(3, 1)` | Insere dimensão depois: uma coluna |


In [17]:
a_b = np.array([10, 20, 30])
print("Original:", a_b.shape)
print("Linha:", a_b[np.newaxis, :].shape)
print("Coluna:", a_b[:, np.newaxis].shape)

Original: (3,)
Linha: (1, 3)
Coluna: (3, 1)


### Caso 8 — o cálculo funciona, mas a intenção está errada

Em uma matriz quadrada `(2, 2)`, somar `(2,)` funciona. Porém, o ajuste é aplicado **por coluna**. Se você pretendia aplicar por linha, precisará de `(2, 1)`.

**Conferir apenas se houve erro não basta. Confira a regra de negócio e o resultado à mão.**


In [18]:
quadrada_b = np.array([[10, 20], [30, 40]])
ajuste_b = np.array([1, 2])
print("Por coluna:\n", quadrada_b + ajuste_b)
print("Por linha:\n", quadrada_b + ajuste_b.reshape(2, 1))

Por coluna:
 [[11 22]
 [31 42]]
Por linha:
 [[11 21]
 [32 42]]


Por coluna: `[[11, 22], [31, 42]]`. Por linha: `[[11, 21], [32, 42]]`. Os resultados são diferentes, embora ambas as operações sejam válidas.

### Checklist para resolver broadcasting

1. O que representa cada linha e cada coluna?
2. O ajuste pertence a todas as posições, a cada coluna ou a cada linha?
3. Quais são os shapes?
4. Os pares de dimensões são iguais ou contêm tamanho 1?
5. Preciso de `reshape`, `keepdims=True` ou `np.newaxis`?
6. O primeiro resultado confere com uma conta feita à mão?

Broadcasting evita criar manualmente arrays repetidos, mas o resultado de uma operação como `a + b` ainda ocupa memória. Uma combinação de muitos valores pode produzir um array grande.


## 7. Combinar condições

Use `&` para **E**, `|` para **OU** e `~` para inverter uma máscara. **Coloque cada comparação entre parênteses.**

Não use `and` ou `or` para combinar esses arrays de condições: queremos comparar cada posição, não decidir um único valor lógico para o array inteiro.

In [19]:
idades = np.array([18, 25, 30, 40, 60])
mascara_intervalo = (idades >= 25) & (idades <= 40)
print("Entre 25 e 40:", idades[mascara_intervalo])
print("Abaixo de 25 OU acima de 40:", idades[(idades < 25) | (idades > 40)])
print("Fora do intervalo:", idades[~mascara_intervalo])

Entre 25 e 40: [25 30 40]
Abaixo de 25 OU acima de 40: [18 60]
Fora do intervalo: [18 60]


O intervalo inclui `[25 30 40]`. As outras duas seleções devolvem `[18 60]`. `~` aplicado à máscara troca True por False e vice-versa.

## 8. any, all e contagem de condições

`np.any(mascara)` responde se **pelo menos uma** posição é True. `np.all(mascara)` responde se **todas** são True. `np.count_nonzero(mascara)` conta as posições True numa máscara booleana.

In [20]:
estoque = np.array([3, 12, 0, 8])
precisa_repor = estoque < 5
print("Máscara:", precisa_repor)
print("Algum precisa repor?", np.any(precisa_repor))
print("Todos precisam repor?", np.all(precisa_repor))
print("Quantos precisam repor?", np.count_nonzero(precisa_repor))

Máscara: [ True False  True False]
Algum precisa repor? True
Todos precisam repor? False
Quantos precisam repor? 2


Há duas posições abaixo de 5: estoques 3 e 0. any é True, all é False e a contagem é 2.

## 9. Tipos e conversão com astype

`dtype` informa o tipo dos elementos. `.astype(tipo)` devolve um array convertido. Antes de converter texto para número, os textos precisam ter formato numérico válido.

Converter números decimais para inteiros remove a parte fracionária, aproximando em direção a zero; isso **não é arredondamento**.

In [21]:
textos = np.array(["10", "20", "30"])
numeros = textos.astype(float)
print("Convertidos:", numeros)
print("Soma:", numeros.sum())

decimais = np.array([2.9, -2.9])
print("Para inteiro:", decimais.astype(int))

Convertidos: [10. 20. 30.]
Soma: 60.0
Para inteiro: [ 2 -2]


Os textos viram `[10. 20. 30.]`, cuja soma é 60.0. Os decimais viram `[2 -2]`. Texto como `"R$ 10"` exige limpeza antes da conversão.

## 10. Valores ausentes: NaN

`np.nan` representa um valor numérico ausente. Use `np.isnan` para identificá-lo: comparar com `== np.nan` não funciona como um teste de ausência.

`np.mean` propaga NaN neste exemplo. `np.nanmean` calcula a média ignorando os ausentes. Escolher ignorá-los depende da regra da análise: um dado faltante não significa zero.

In [22]:
medicoes = np.array([10.0, np.nan, 20.0])
print("Ausentes:", np.isnan(medicoes))
print("Quantidade:", np.count_nonzero(np.isnan(medicoes)))
print("Média comum:", np.mean(medicoes))
print("Média sem ausentes:", np.nanmean(medicoes))
print("Somente preenchidos:", medicoes[~np.isnan(medicoes)])

Ausentes: [False  True False]
Quantidade: 1
Média comum: nan
Média sem ausentes: 15.0
Somente preenchidos: [10. 20.]


A média comum é nan. A média dos valores disponíveis é (10 + 20) ÷ 2 = 15.0. Se não houver valores disponíveis, nanmean não terá uma média válida e pode emitir um aviso.

## 11. Unir arrays e encontrar valores únicos

`np.concatenate` une arrays ao longo de um eixo existente. Na matriz, `axis=0` acrescenta linhas e `axis=1` acrescenta colunas; os tamanhos das demais dimensões precisam combinar.

`np.unique(..., return_counts=True)` devolve dois resultados: os valores únicos (ordenados por padrão) e suas contagens. Podemos guardar os dois com `valores, contagens = ...`.

In [23]:
grupo_a = np.array([10, 20])
grupo_b = np.array([30, 40])
print("Unidos:", np.concatenate([grupo_a, grupo_b]))

categorias = np.array([2, 1, 2, 3, 1, 2])
unicos, contagens = np.unique(categorias, return_counts=True)
print("Valores:", unicos)
print("Contagens:", contagens)

Unidos: [10 20 30 40]
Valores: [1 2 3]
Contagens: [2 3 1]


A união é `[10 20 30 40]`. Os valores únicos são `[1 2 3]` e as contagens correspondentes são `[2 3 1]`. A segunda contagem 3 pertence ao valor 2.

## 12. Aleatoriedade reproduzível

`np.random.default_rng(42)` cria um gerador com uma **semente**. Um novo gerador com a mesma semente reproduz a sequência no mesmo ambiente. Isso ajuda a repetir exercícios.

`rng.integers(1, 7, size=5)` sorteia cinco inteiros de 1 a 6: o limite superior 7 não entra. Chamadas sucessivas avançam o estado do gerador e produzem novos sorteios.

In [24]:
rng = np.random.default_rng(42)
sorteios = rng.integers(1, 7, size=5)
print("Sorteios:", sorteios)

rng_repetido = np.random.default_rng(42)
print("Novo gerador, mesma semente:", rng_repetido.integers(1, 7, size=5))

Sorteios: [1 5 4 3 3]
Novo gerador, mesma semente: [1 5 4 3 3]


As duas saídas coincidem porque criamos dois geradores com a mesma semente. Sortear novamente no primeiro gerador avança a sequência.

## 13. Exercícios — tente antes de consultar as resoluções

Em cada exercício, execute a preparação, preveja o resultado e escreva sua tentativa na célula seguinte. Os exercícios 1 a 5 trabalham com matrizes; 6 a 10, seleção e resumo; 11 a 15, aplicações e integração.


### Exercício 1 — Localizar um valor

**Pedido:** Mostre o valor da segunda linha, terceira coluna.

**Resultado esperado:** 12

In [25]:
m1 = np.array([[2, 4, 6], [8, 10, 12]])

In [26]:
# Escreva sua tentativa aqui.


### Exercício 2 — Selecionar uma coluna

**Pedido:** Selecione a segunda coluna de todas as linhas.

**Resultado esperado:** [10 20 30]

In [27]:
m2 = np.array([[5, 10], [15, 20], [25, 30]])

In [28]:
# Escreva sua tentativa aqui.


### Exercício 3 — Total por loja

**Pedido:** Cada linha é uma loja. Calcule o total de cada loja.

**Resultado esperado:** [30 70 110]

In [29]:
m3 = np.array([[10, 20], [30, 40], [50, 60]])

In [30]:
# Escreva sua tentativa aqui.


### Exercício 4 — Média por dia

**Pedido:** As linhas são lojas e as colunas são dias. Calcule a média das lojas para cada dia.

**Resultado esperado:** [20. 30. 40.]

In [31]:
m4 = np.array([[10, 20, 30], [30, 40, 50]])

In [32]:
# Escreva sua tentativa aqui.


### Exercício 5 — Reorganizar uma sequência

**Pedido:** Crie os números de 1 a 12 e organize-os em três linhas e quatro colunas.

**Resultado esperado:** [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]

In [33]:
# Escreva sua tentativa aqui.


### Exercício 6 — Filtrar um intervalo

**Pedido:** Selecione valores entre 20 e 50, incluindo os limites.

**Resultado esperado:** [20 35 50]

In [34]:
a6 = np.array([10, 20, 35, 50, 70])

In [35]:
# Escreva sua tentativa aqui.


### Exercício 7 — Selecionar extremos

**Pedido:** Selecione valores menores que 10 OU maiores que 50.

**Resultado esperado:** [5 60]

In [36]:
a7 = np.array([5, 10, 25, 50, 60])

In [37]:
# Escreva sua tentativa aqui.


### Exercício 8 — Verificar estoques

**Pedido:** Informe se todos os estoques são positivos e quantos são zero.

**Resultado esperado:** Todos positivos: False; zeros: 2

In [38]:
a8 = np.array([3, 0, 8, 0])

In [39]:
# Escreva sua tentativa aqui.


### Exercício 9 — Converter números em texto

**Pedido:** Converta os textos para float e calcule a soma.

**Resultado esperado:** 30.0

In [40]:
a9 = np.array(["12.5", "7.5", "10"])

In [41]:
# Escreva sua tentativa aqui.


### Exercício 10 — Trabalhar com ausência

**Pedido:** Conte os ausentes e calcule a média dos valores disponíveis.

**Resultado esperado:** Ausentes: 2; média: 10.0

In [42]:
a10 = np.array([5.0, np.nan, 15.0, np.nan])

In [43]:
# Escreva sua tentativa aqui.


### Exercício 11 — Aplicar bônus por coluna

**Pedido:** Cada coluna é um dia. Some os bônus correspondentes aos valores das duas lojas.

**Resultado esperado:** [[12, 24, 36], [42, 54, 66]]

In [44]:
m11 = np.array([[10, 20, 30], [40, 50, 60]])
bonus11 = np.array([2, 4, 6])

In [45]:
# Escreva sua tentativa aqui.


### Exercício 12 — Unir matrizes por linhas

**Pedido:** Acrescente a linha de uma terceira loja abaixo das duas primeiras.

**Resultado esperado:** [[10, 20], [30, 40], [50, 60]]

In [46]:
lojas12 = np.array([[10, 20], [30, 40]])
nova12 = np.array([[50, 60]])

In [47]:
# Escreva sua tentativa aqui.


### Exercício 13 — Contar categorias

**Pedido:** Mostre cada código único e sua frequência.

**Resultado esperado:** Códigos: [1 2 3]; contagens: [3 2 1]

In [48]:
codigos13 = np.array([1, 2, 1, 3, 2, 1])

In [49]:
# Escreva sua tentativa aqui.


### Exercício 14 — Sortear dados

**Pedido:** Crie um gerador com semente 7 e sorteie seis inteiros de 1 a 6.

**Resultado esperado:** Array com seis inteiros entre 1 e 6, reproduzível ao recriar o gerador com semente 7

In [50]:
# Escreva sua tentativa aqui.


### Exercício 15 — Desafio: relatório de lojas

**Pedido:** Cada linha é uma loja e cada coluna é um dia. Calcule os totais por loja e por dia. Selecione as linhas das lojas cujo total é >= 100. Rotule cada loja com np.where: total >= 100 recebe "Meta atingida"; caso contrário, "Abaixo da meta".

**Resultado esperado:** Totais por loja: [90 180 60]; por dia: [80 110 140]; somente a segunda loja atinge a meta

In [51]:
m15 = np.array([[20, 30, 40], [50, 60, 70], [10, 20, 30]])

In [52]:
# Escreva sua tentativa aqui.


## 13B. Oficina de broadcasting — 8 exercícios extras

Para cada exercício: identifique os eixos, anote os formatos, faça uma conta à mão e só depois escreva o código. As resoluções ficam após as 15 resoluções gerais.

### B1 — Escalar

**Pedido:** Some 10 a todos os elementos.

**Esperado:** [[11, 12], [13, 14]]

In [53]:
b1 = np.array([[1, 2], [3, 4]])

In [54]:
# Sua tentativa:


### B2 — Por coluna

**Pedido:** Some os ajustes correspondentes a cada coluna.

**Esperado:** [[12, 24, 36], [42, 54, 66]]

In [55]:
b2 = np.array([[10, 20, 30], [40, 50, 60]])
aj2 = np.array([2, 4, 6])

In [56]:
# Sua tentativa:


### B3 — Por linha

**Pedido:** Aplique 5 à primeira linha e 10 à segunda. Prepare o ajuste no formato correto.

**Esperado:** [[15, 25, 35], [50, 60, 70]]

In [57]:
b3 = np.array([[10, 20, 30], [40, 50, 60]])
aj3 = np.array([5, 10])

In [58]:
# Sua tentativa:


### B4 — Prever formatos

**Pedido:** Sem tentar somar, explique quais formatos podem ser combinados com a matriz abaixo: (3,), (2, 1), (2,) e (1, 3).

**Esperado:** (3,), (2, 1) e (1, 3) funcionam; (2,) não

In [59]:
b4 = np.zeros((2, 3))

In [60]:
# Sua tentativa:


### B5 — Combinações

**Pedido:** Crie uma matriz com cada frete combinado com cada preço.

**Esperado:** [[25, 45, 65], [30, 50, 70]]

In [61]:
fretes5 = np.array([5, 10])
precos5 = np.array([20, 40, 60])

In [62]:
# Sua tentativa:


### B6 — Média por linha

**Pedido:** Subtraia a média de cada linha, preservando a dimensão da média com keepdims=True.

**Esperado:** [[-2, 0, 2], [-10, 0, 10]]

In [63]:
b6 = np.array([[2, 4, 6], [10, 20, 30]])

In [64]:
# Sua tentativa:


### B7 — Desconto por produto

**Pedido:** As colunas são produtos. Aplique 10% de desconto ao primeiro e 20% ao segundo.

**Esperado:** [[90, 160], [45, 64]]

In [65]:
b7 = np.array([[100, 200], [50, 80]])

In [66]:
# Sua tentativa:


### B8 — Conferir a intenção

**Pedido:** A primeira filial deve receber aumento de 10% e a segunda de 20%. As linhas são filiais. Faça o cálculo e explique por que multiplicar diretamente por fatores8 aplicaria o aumento por produto.

**Esperado:** [[110, 220], [360, 480]]

In [67]:
b8 = np.array([[100, 200], [300, 400]])
fatores8 = np.array([1.10, 1.20])

In [68]:
# Sua tentativa:


## 14. Resoluções comentadas

Confira as contas e o significado dos eixos, não apenas a aparência da saída.

### Resolução 1 — Localizar um valor

Os índices da segunda linha e da terceira coluna são 1 e 2.

In [69]:
m1 = np.array([[2, 4, 6], [8, 10, 12]])

print(m1[1, 2])

12


### Resolução 2 — Selecionar uma coluna

: seleciona todas as linhas; 1 seleciona a segunda coluna.

In [70]:
m2 = np.array([[5, 10], [15, 20], [25, 30]])

print(m2[:, 1])

[10 20 30]


### Resolução 3 — Total por loja

axis=1 reúne as colunas de cada linha; o resultado tem um total por loja.

In [71]:
m3 = np.array([[10, 20], [30, 40], [50, 60]])

print(m3.sum(axis=1))

[ 30  70 110]


### Resolução 4 — Média por dia

axis=0 reúne os valores das linhas para cada coluna: (10+30)/2 e assim por diante.

In [72]:
m4 = np.array([[10, 20, 30], [30, 40, 50]])

print(m4.mean(axis=0))

[20. 30. 40.]


### Resolução 5 — Reorganizar uma sequência

13 fica fora do arange. Os 12 elementos cabem em 3 × 4.

In [73]:
seq5 = np.arange(1, 13)
print(seq5.reshape(3, 4))

[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]


### Resolução 6 — Filtrar um intervalo

Cada comparação tem parênteses. & exige que ambas sejam verdadeiras.

In [74]:
a6 = np.array([10, 20, 35, 50, 70])

print(a6[(a6 >= 20) & (a6 <= 50)])

[20 35 50]


### Resolução 7 — Selecionar extremos

| permite que pelo menos uma condição seja verdadeira. 10 e 50 não entram.

In [75]:
a7 = np.array([5, 10, 25, 50, 60])

print(a7[(a7 < 10) | (a7 > 50)])

[ 5 60]


### Resolução 8 — Verificar estoques

all verifica todas as posições; count_nonzero conta True na comparação com zero.

In [76]:
a8 = np.array([3, 0, 8, 0])

print("Todos positivos:", np.all(a8 > 0))
print("Zeros:", np.count_nonzero(a8 == 0))

Todos positivos: False
Zeros: 2


### Resolução 9 — Converter números em texto

Os textos usam ponto decimal e podem ser convertidos diretamente. A soma é 12.5 + 7.5 + 10.

In [77]:
a9 = np.array(["12.5", "7.5", "10"])

numeros9 = a9.astype(float)
print(numeros9.sum())

30.0


### Resolução 10 — Trabalhar com ausência

isnan identifica duas ausências. nanmean calcula (5+15)/2 sem tratá-las como zero.

In [78]:
a10 = np.array([5.0, np.nan, 15.0, np.nan])

print("Ausentes:", np.count_nonzero(np.isnan(a10)))
print("Média:", np.nanmean(a10))

Ausentes: 2
Média: 10.0


### Resolução 11 — Aplicar bônus por coluna

O formato (3,) se alinha às três colunas de (2, 3). O bônus é aplicado em cada linha.

In [79]:
m11 = np.array([[10, 20, 30], [40, 50, 60]])
bonus11 = np.array([2, 4, 6])

print(m11 + bonus11)

[[12 24 36]
 [42 54 66]]


### Resolução 12 — Unir matrizes por linhas

axis=0 acrescenta linhas. As duas matrizes têm duas colunas; a nova linha foi preparada com formato (1, 2).

In [80]:
lojas12 = np.array([[10, 20], [30, 40]])
nova12 = np.array([[50, 60]])

unidas12 = np.concatenate([lojas12, nova12], axis=0)
print(unidas12)

[[10 20]
 [30 40]
 [50 60]]


### Resolução 13 — Contar categorias

A atribuição recebe os dois resultados. As posições dos códigos e das contagens correspondem entre si.

In [81]:
codigos13 = np.array([1, 2, 1, 3, 2, 1])

valores13, contagens13 = np.unique(codigos13, return_counts=True)
print("Códigos:", valores13)
print("Contagens:", contagens13)

Códigos: [1 2 3]
Contagens: [3 2 1]


### Resolução 14 — Sortear dados

O limite superior 7 é excluído. size=6 pede seis resultados. A semente permite repetir a sequência no mesmo ambiente.

In [82]:
rng14 = np.random.default_rng(7)
print(rng14.integers(1, 7, size=6))

[6 4 5 6 4 5]


### Resolução 15 — Desafio: relatório de lojas

A máscara tem um booleano por loja: False, True, False. Usada na primeira dimensão de m15, seleciona a linha inteira da segunda loja.

In [83]:
m15 = np.array([[20, 30, 40], [50, 60, 70], [10, 20, 30]])

totais15 = m15.sum(axis=1)
print("Por loja:", totais15)
print("Por dia:", m15.sum(axis=0))
mascara15 = totais15 >= 100
print("Lojas selecionadas:\n", m15[mascara15])
print("Situação:", np.where(mascara15, "Meta atingida", "Abaixo da meta"))

Por loja: [ 90 180  60]
Por dia: [ 80 110 140]
Lojas selecionadas:
 [[50 60 70]]
Situação: ['Abaixo da meta' 'Meta atingida' 'Abaixo da meta']


## 14B. Resoluções da oficina de broadcasting


### B1 — Escalar

O escalar se aplica a todas as posições.

In [84]:
b1 = np.array([[1, 2], [3, 4]])

print(b1 + 10)

[[11 12]
 [13 14]]


### B2 — Por coluna

(3,) se alinha à última dimensão de (2, 3).

In [85]:
b2 = np.array([[10, 20, 30], [40, 50, 60]])
aj2 = np.array([2, 4, 6])

print(b2 + aj2)

[[12 24 36]
 [42 54 66]]


### B3 — Por linha

(2, 1) combina com (2, 3). O tamanho 1 permite usar cada ajuste nas três colunas.

In [86]:
b3 = np.array([[10, 20, 30], [40, 50, 60]])
aj3 = np.array([5, 10])

col3 = aj3.reshape(2, 1)
print(b3 + col3)

[[15 25 35]
 [50 60 70]]


### B4 — Prever formatos

np.zeros cria a matriz de zeros para observar o formato. Compare os shapes da direita para a esquerda.

In [87]:
b4 = np.zeros((2, 3))

print("Formato da base:", b4.shape)
print("Compatíveis: (3,), (2, 1), (1, 3)")
print("Incompatível: (2,), pois a dimensão final 2 conflita com 3")

Formato da base: (2, 3)
Compatíveis: (3,), (2, 1), (1, 3)
Incompatível: (2,), pois a dimensão final 2 conflita com 3


### B5 — Combinações

Uma coluna de dois fretes e uma linha de três preços produzem todas as seis combinações.

In [88]:
fretes5 = np.array([5, 10])
precos5 = np.array([20, 40, 60])

print(fretes5.reshape(2, 1) + precos5.reshape(1, 3))

[[25 45 65]
 [30 50 70]]


### B6 — Média por linha

As médias 4 e 20 ficam no formato (2, 1), próprio para subtrair de cada linha.

In [89]:
b6 = np.array([[2, 4, 6], [10, 20, 30]])

med6 = b6.mean(axis=1, keepdims=True)
print(b6 - med6)

[[ -2.   0.   2.]
 [-10.   0.  10.]]


### B7 — Desconto por produto

Cada fator se aplica à coluna correspondente. As duas linhas recebem os mesmos descontos por produto.

In [90]:
b7 = np.array([[100, 200], [50, 80]])

fatores7 = np.array([0.90, 0.80])
print(b7 * fatores7)

[[ 90. 160.]
 [ 45.  64.]]


### B8 — Conferir a intenção

Em (2, 2) com (2,), os fatores alinham com as colunas. reshape(2, 1) expressa a intenção por linha. Podem aparecer pequenas diferenças de representação decimal na impressão.

In [91]:
b8 = np.array([[100, 200], [300, 400]])
fatores8 = np.array([1.10, 1.20])

print("Por filial:\n", b8 * fatores8.reshape(2, 1))
print("Direto aplica por coluna:\n", b8 * fatores8)

Por filial:
 [[110. 220.]
 [360. 480.]]
Direto aplica por coluna:
 [[110. 240.]
 [330. 480.]]


## 15. Resumão e erros comuns

| Quero... | Comando |
|---|---|
| Selecionar coluna | `m[:, 1]` |
| Somar por linha | `m.sum(axis=1)` |
| Somar por coluna | `m.sum(axis=0)` |
| Reorganizar | `a.reshape(2, 3)` |
| Transpor | `m.T` |
| Voltar a uma dimensão com cópia | `m.flatten()` |
| Combinar condições com E | `(a >= 10) & (a <= 20)` |
| Combinar condições com OU | `(a < 10) | (a > 20)` |
| Verificar algum / todos | `np.any(mascara)` / `np.all(mascara)` |
| Contar True | `np.count_nonzero(mascara)` |
| Converter tipo | `a.astype(float)` |
| Identificar ausentes | `np.isnan(a)` |
| Média ignorando ausentes | `np.nanmean(a)` |
| Unir arrays | `np.concatenate([a, b])` |
| Valores únicos e contagens | `np.unique(a, return_counts=True)` |
| Criar gerador aleatório | `np.random.default_rng(42)` |

**Se der erro:** confira o formato com `.shape`, a quantidade com `.size` e o tipo com `.dtype`.

- reshape exige o mesmo total de elementos.
- Broadcasting exige dimensões finais iguais ou de tamanho 1.
- Filtros combinados precisam de parênteses e operadores `&` e `|`.
- Ausência não é zero: escolha a regra de tratamento antes do cálculo.
- Transposição, reshape e fatiamento podem compartilhar dados. Use `.copy()` quando precisar editar de forma independente.

**Para verificar seu entendimento:** explique por que `axis=1` entrega um total por loja na nossa tabela e por que `axis=0` entrega um total por dia.
